src/routes/version_creation.rs

```diff
@@ -0,0 +1,239 @@
+        } else {
+            return Err(CreateError::MissingValueError(
+                "Missing content file extension".to_string(),
+            ));
+        };
+
+        if &*file_extension == "jar" {
+            let version = version_builder.as_mut().ok_or_else(|| {
+                CreateError::InvalidInput(String::from("`data` field must come before file fields"))
+            })?;
+
+            let mut data = Vec::new();
+            while let Some(chunk) = field.next().await {
+                data.extend_from_slice(&chunk.map_err(CreateError::MultipartError)?);
+            }
+
+            let upload_data = file_host
+                .upload_file(
+                    "application/java-archive",
+                    &format!(
+                        "{}/{}/{}",
+                        version.mod_id.0, version.version_number, file_name
+                    ),
+                    data.to_vec(),
+                )
+                .await?;
+
+            uploaded_files.push(UploadedFile {
+                file_id: upload_data.file_id.clone(),
+                file_name: upload_data.file_name.clone(),
+            });
+
+            // Add the newly uploaded file to the existing or new version
+
+            // TODO: Malware scan + file validation
+            version
+                .files
+                .push(models::version_item::VersionFileBuilder {
+                    filename: file_name.to_string(),
+                    url: format!("{}/{}", cdn_url, upload_data.file_name),
+                    hashes: vec![models::version_item::HashBuilder {
+                        algorithm: "sha1".to_string(),
+                        // This is an invalid cast - the database expects the hash's
+                        // bytes, but this is the string version.
+                        hash: upload_data.content_sha1.into_bytes(),
+                    }],
+                });
+        }
+    }
+
+    let version_data_safe = initial_version_data
+        .ok_or_else(|| CreateError::InvalidInput("`data` field is required".to_string()))?;
+    let version_builder_safe = version_builder
+        .ok_or_else(|| CreateError::InvalidInput("`data` field is required".to_string()))?;
+
+    let response = Version {
+        id: VersionId(version_builder_safe.version_id.0 as u64),
+        mod_id: ModId(version_builder_safe.mod_id.0 as u64),
+        name: version_builder_safe.name.clone(),
+        version_number: version_builder_safe.version_number.clone(),
+        changelog_url: version_builder_safe.changelog_url.clone(),
+        date_published: chrono::Utc::now(),
+        downloads: 0,
+        version_type: version_data_safe.release_channel,
+        files: version_builder_safe
+            .files
+            .iter()
+            .map(|_x| VersionFile {
+                // TODO: Put real data here
+                hashes: vec![],
+                url: String::new(),
+            })
```

Comment:
I believe this has all of the data available if this is returning `VersionFile`s instead of `FileId`s:
```rust
.map(|file| VersionFile {
    hashes: file.hashes.iter().map(|hash| FileHash {
        algorithm: hash.algorithm,
        // This is a hack since the hashes are currently stored as ASCII
        // in the database, but represented here as a Vec<u8>.  At some
        // point we need to change the hash to be the real bytes  in the
        // database and add more processing here.
        hash: String::from_utf8(hash.hash).unwrap(),
    }).collect(),
    url: file.url,
})
```